datalad/core/distributed/clone.py

```diff
@@ -0,0 +1,625 @@
+        except OSError as e:
+            yield get_status_dict(
+                status='error',
+                path=path,
+                message=('cannot handle target path: %s', exc_str(e)),
+                **result_props)
+            return
+
+        destination_dataset = Dataset(path)
+        result_props['ds'] = destination_dataset
+
+        if ds is not None and ds.pathobj not in path.parents:
+            yield get_status_dict(
+                status='error',
+                message=("clone target path '%s' not in specified target dataset '%s'",
+                         path, ds),
+                **result_props)
+            return
+
+        # perform the actual cloning operation
+        yield from clone_dataset(
+            [source],
+            destination_dataset,
+            reckless,
+            description,
+            result_props,
+        )
+
+        if ds is not None:
+            # we created a dataset in another dataset
+            # -> make submodule
+            for r in ds.save(
+                    path,
+                    return_type='generator',
+                    result_filter=None,
+                    result_xfm=None,
+                    on_failure='ignore'):
+                yield r
+
+
+def clone_dataset(
+        srcs,
+        destds,
+        reckless=None,
+        description=None,
+        result_props=None):
+    """Internal helper to perform cloning without sanity checks (assumed done)
+
+    This helper does not handle any saving of subdataset modification or adding
+    in a superdataset.
+
+    Parameters
+    ----------
+    srcs : list
+      Any suitable clone source specifications (paths, URLs)
+    destds : Dataset
+      Dataset instance for the clone destination
+    reckless : {None, 'auto'}, optional
+      Mode switch to put cloned dataset into throw-away configurations, i.e.
+      sacrifice data safety for performance or reesource footprint.
+    description : str, optional
+      Location description for the annex of the dataset clone (if there is any).
+    result_props : dict, optional
+      Default properties for any yielded result, pass on to get_status_dict().
+
+    Yields
+    ------
+    dict
+      DataLad result records
+    """
+    if not result_props:
+        # in case the caller had no specific idea on how results should look
+        # like, provide sensible defaults
+        result_props = dict(
+            action='install',
+            logger=lgr,
+            ds=destds,
+        )
+
+    dest_path = destds.pathobj
+
+    # generate candidate URLs from source argument to overcome a few corner cases
+    # and hopefully be more robust than git clone
+    candidate_sources = [
+        s for src in srcs
+        for s in _get_flexible_source_candidates(src)
+    ]
+
+    # important test! based on this `rmtree` will happen below after failed clone
+    dest_path_existed = dest_path.exists()
+    if dest_path_existed and any(dest_path.iterdir()):
```

Comment:
Going through this entire PR, this line makes me wonder. As far as I can tell, we never tested destination path to be a directory. If it isn't, then how does that line fail? `AttributeError` (no attribute `iterdir`)? Guess it could simply be included in sanity checks within `Clone.__call__`.

Reply:
Although, thinking about it. It probably should be tested by `Dataset` or `*Repo` constructors. If path exists, it has to be a directory.

Reply:
In this context `dest_path` is the `.pathobj` property of a `Dataset` instance. I agree that it is out-of-scope for `clone` to test whether it got a "file dataset".